core/shared/src/main/scala/org/http4s/internal/Logger.scala

```diff
@@ -37,27 +39,40 @@ object Logger {
       message.headers.mkString("Headers(", ", ", ")", redactHeadersWhen)
     else ""
 
-  def defaultLogBody[F[_]: Concurrent](
+  def defaultLogBody[F[_]](
       message: Message[F]
-  )(logBody: Boolean): Option[F[String]] =
+  )(logBody: Boolean)(implicit F: Async[F]): Option[F[String]] =
     if (logBody) {
       val isBinary = message.contentType.exists(_.mediaType.binary)
       val isJson = message.contentType.exists(mT =>
         mT.mediaType == MediaType.application.json || mT.mediaType.subType.endsWith("+json")
       )
-      val bodyStream = if (!isBinary || isJson) {
-        message.bodyText
-      } else {
-        message.body.map(b => java.lang.Integer.toHexString(b & 0xff))
+
+      message.entity match {
+        case Entity.Empty =>
+          None
+        case Entity.Strict(bv) =>
+          if (!isBinary || isJson) {
+            val charset = message.charset.fold(Charset.`UTF-8`.nioCharset)(_.nioCharset)
+            Some(F.delay(bv.decodeStringLenient()(charset)))
+          } else
+            Some(F.delay(bv.toHex))
+        case Entity.Default(_, _) =>
+          val stream =
+            if (!isBinary || isJson)
+              message.bodyText
+            else
+              message.body.map(b => java.lang.Integer.toHexString(b & 0xff))
+
+          Some(stream.compile.string)
       }
-      Some(bodyStream.compile.string)
     } else None
 
   def logMessage[F[_]](message: Message[F])(
       logHeaders: Boolean,
       logBody: Boolean,
       redactHeadersWhen: CIString => Boolean = Headers.SensitiveHeaders.contains,
-  )(log: String => F[Unit])(implicit F: Concurrent[F]): F[Unit] = {
+  )(log: String => F[Unit])(implicit F: Async[F]): F[Unit] = {
```

Comment:
Why does this become `Async`? Why do we need `delay` ?

Reply:
> Why does this become Async?

Because of `delay`. 

> Why do we need delay ?

It _seems_ like we want to suspend the execution of this effect. Generally, it may not be pure, especially since we give away the responsibility of decoding to a third-party library (even if it's astonishing fs2). And we're not using `blocking` since we believe fs2's decoding isn't thread-blocking.